tensorflow/lite/kernels/internal/common.h

```diff
@@ -181,46 +182,46 @@ inline int32_t MultiplyByQuantizedMultiplier(int64_t x,
   assert(x >= -(static_cast<int64_t>(1) << 47) &&
          x < (static_cast<int64_t>(1) << 47));
 
-  int32_t reduced_multiplier = (quantized_multiplier < 0x7FFF0000)
-                                   ? ((quantized_multiplier + (1 << 15)) >> 16)
-                                   : 0x7FFF;
-  int total_shift = 15 - shift;
-  x = (x * (int64_t)reduced_multiplier) + ((int64_t)1 << (total_shift - 1));
-  int32_t result = x >> total_shift;
-  return result;
+  const int32_t reduced_multiplier =
+      (quantized_multiplier < 0x7FFF0000)
+          ? ((quantized_multiplier + (1 << 15)) >> 16)
+          : 0x7FFF;
+  const int64_t total_shift = 15 - shift;
+  const int64_t round = static_cast<int64_t>(1) << (total_shift - 1);
+  int64_t result = x * static_cast<int64_t>(reduced_multiplier) + round;
+  result = result >> total_shift;
+
+  assert(result >= std::numeric_limits<int32_t>::min() &&
+         result <= std::numeric_limits<int32_t>::max());
+  return static_cast<int32_t>(result);
 }
 
 #ifdef USE_NEON
-// Round uses ARM's rounding shift right.
 inline int32x4x4_t MultiplyByQuantizedMultiplier4Rows(
     int32x4x4_t input_val, int32_t quantized_multiplier, int shift) {
-  const int left_shift = std::max(shift, 0);
-  const int right_shift = std::min(shift, 0);
+  const int right_shift = std::min(-1, shift);
+  const int left_shift = shift - right_shift;
+
+  const int32x4_t multiplier_dup = vdupq_n_s32(quantized_multiplier);
+  const int32x4_t left_shift_dup = vdupq_n_s32(left_shift);
+  const int32x4_t right_shift_dup = vdupq_n_s32(right_shift);
+
   int32x4x4_t result;
+  result.val[0] = vrshlq_s32(
+      vqdmulhq_s32(vshlq_s32(input_val.val[0], left_shift_dup), multiplier_dup),
+      right_shift_dup);
+
+  result.val[1] = vrshlq_s32(
+      vqdmulhq_s32(vshlq_s32(input_val.val[1], left_shift_dup), multiplier_dup),
+      right_shift_dup);
+
+  result.val[2] = vrshlq_s32(
+      vqdmulhq_s32(vshlq_s32(input_val.val[2], left_shift_dup), multiplier_dup),
```

Comment:
why change this?

vqrdmulhq_s32 is rounding right?

Reply:
Yes, vqrdmulhq_s32 is a multiplication with rounding which we replace with vqdmulhq_s32 which is a multiplication with truncation to avoid the double-rounding (with vrshlq_s32 as second rounding). 

This change is aligned to the change done in https://github.com/google/ruy/pull/227 (the discussion there provide more details) and matches the behaviour of the modified MultiplyByQuantizedMultiplier .

Reply:
thanks for the explanation!